src/Simulation/QCTraceSimulator/SortedQubitPool.cs

```diff
@@ -0,0 +1,321 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.Quantum.Simulation.QCTraceSimulatorRuntime {
+
+    /// <summary>
+    /// Time to track qubit availability (DepthTime). Time is also used to track dependencies:
+    /// Gates that happen later on a qubit cannot be scheduled before gates on the same qubit.
+    /// This is tracked transitively via time, but zero gate depth prevents this from happening.
+    /// This is solved by having additional time component - DependencyTime.
+    /// </summary>
+    internal struct ComplexTime
+    {
+        /// <summary>
+        /// Time used to track qubit availability time. Double to match configuration interface.
+        /// </summary>
+        internal double DepthTime { get; private set; }
+
+        /// <summary>
+        /// This is only used for comparisons when DepthTime is equal. This time tracks dependencies.
+        /// </summary>
+        private long DependencyTime;
+
+        internal ComplexTime(double time) {
+            DepthTime = time;
+            DependencyTime = 0;
+        }
+
+        private ComplexTime(double depthTime, long dependencyTime) {
+            DepthTime = depthTime;
+            DependencyTime = dependencyTime;
+        }
+
+        public override string ToString() {
+            return $"{DepthTime}d+{DependencyTime}";
+        }
+        /// <summary>
+        /// Add gate time to availability time. If gate time is zero, advance dependency time.
+        /// </summary>
+        /// <param name="time">Gate time to advance by. Assumed to be precise for comparison with 0.</param>
+        /// <returns>ComplexTime advanced by provided gate time</returns>
+        internal ComplexTime AdvanceBy(double time) {
+            if (time == 0.0) {
+                return new ComplexTime(DepthTime, DependencyTime + 1);
+            }
+            return new ComplexTime(DepthTime + time);
+        }
+
+        /// <summary>
+        /// Subtracts argument ComplexTime from "this" ComplexTime. Returns the result.
+        /// Depth times are assumed to be precise  for comparison.
+        /// </summary>
+        /// <param name="time">Time to subtract from this object.</param>
+        /// <returns>Result of subtraction.</returns>
+        internal ComplexTime Subtract(ComplexTime time) {
+            if (DepthTime < time.DepthTime) {
+                throw new ArgumentException("Result of ComplexTime subtraction is negative.");
+            }
+            if (DepthTime == time.DepthTime) {
+                return new ComplexTime(0, DependencyTime - time.DependencyTime);
```

Comment:
Is it OK for the DependencyTime to be negative?

Reply:
No. We should always subtract smaller value from larger value. I do checks for DepthTime because it is double and floating point glitches may happen. I'll do it a little different to emphasize this.

Reply:
Updated wording will be in the next push.